### PR TITLE
Add XOAUTH2 support

### PR DESCRIPTION
**Pull request by tats**
cf. https://github.com/kazu-yamamoto/Mew/issues/172

### CODE REVIEW

**Comment by tats** [comment]
Updated commit logs to fix incorrect mew-config-alist.

File: elisp/mew-auth.el [hunk 1]
```diff
@@ -67,6 +67,80 @@
...
+  ;; https://developers.google.com/gmail/imap/xoauth2-protocol#error_response_2
+  (require 'json)
+  (let ((json-status
+         (ignore-errors
+           (json-read-from-string
+            (base64-decode-string status-string)))))
+    (if json-status
+        (if (string-match "^2" (cdr (assoc 'status json-status)))
+            "OK" ;; 2XX
+          "NO" ;; XXX: Anyway NO?
```
Comment (kazu-yamamoto) [inline elisp/mew-auth.el:87]: This line shoud be `"OK") ;; 2XX` and the next line should be removed.

File: elisp/mew-auth.el [hunk 2]
```diff
@@ -67,6 +67,80 @@
...
+          )
+      "OK" ;; XXX: Maybe OK if not JSON.
```
Comment (kazu-yamamoto) [inline elisp/mew-auth.el:89]: Ditto.

File: elisp/mew-auth.el (hunk 1 above)
Comment (yoshinari-nomura) [reply elisp/mew-auth.el:87]: Let me clarify my understanding.
"This line" means line 86 or 87?
If 86, we can not have a chance to return "NO" on failure.

Gmail seems not to return JSON string on success.
So, I want to return *"OK"* if `json-string` is nil.
If `json-string` is valid, I want to respect the `status` value in it.

Comment (kazu-yamamoto) [reply elisp/mew-auth.el:87]: I don't talk about logic, but talk about coding-style.
Closing brackets must not consume a line by themselves.
This is *the* custom shared by many Lispers.

Comment (yoshinari-nomura) [reply elisp/mew-auth.el:87]: Ah..., thanks. I'll remove some newlines.

To tell the truth, Line 88  and 89 should have had some code to take care other OAuth providers such as Office 365.
It was the lingering newline of my hesitation :sweat_smile: 

By the way, is there any handy way for me to repair this PR authored by @tats san?
If not, do you mind merging this PR with the unwilling nelines?
Following this PR, I'll make a new PR that removes newlines.

**Comment by tats** [comment]
Rebased and pushed the coding style fixes.

@kazu-yamamoto Your comment seemed requesting removal of "NO", so confused.

@yoshinari-nomura Anyway, thanks for the clarification.
Please tell me if more fix needed.  I'll cherry-pick from your branch.
Instead, you may create a new pull request based on this pull request.

**Comment by kazu-yamamoto** [comment]
I will merge this PR with brackets fixed.

**Comment by kazu-yamamoto** [comment]
@tats already fixed the brackets. So, I merged this PR.
Thank you all for your contribution!

**Comment by yoshinari-nomura** [comment]
@tats @kazu-yamamoto Thanks !!

**Pull request merged**